### PR TITLE
CC-1178 php 7.4 on docker CI

### PR DESCRIPTION
**Pull request by halles**

### CODE REVIEW

File: ci/deploy-to-production [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+cd /var/www
+
+# Creates a deployment request for Production
+# The deployment gets queued and will be waiting for approval
+# from a release manager in CWP's dashboard,
+# as per the `bypass_and_start` parameter in the request.
+curl -X POST -u $CWP_DASH_USER:$CWP_DASH_TOKEN \
+  https://dash.cwp.govt.nz/naut/project/$CWP_STACK_ID/environment/prod/deploys \
+  -H 'Content-Type: application/json' -H 'Accept: application/json' \
+  --data "{ \
```
Comment (G-Rath) [inline ci/deploy-to-production:17]: 🐘 this would be good to have in its own JSON file, which we could use `envsubt` to replace the variables (or `sed`).

File: ci/deploy-to-uat [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+# It will always be "Queued" at first.
+STATUS=$(cat response.json | jq -r '.data.attributes.status')
+
+# We use a while loop to wait a few seconds, and query CWP on the
+# fetch request. If the query has finished, our status will change
+# to Complete. If status is not achieved we loop again.
+while [ "$STATUS" != "Complete" ]; do
+  sleep 15
+  STATUS=$(curl --silent -X GET -u $CWP_DASH_USER:$CWP_DASH_TOKEN $LINK \
+    -H 'Content-Type: application/json' -H 'Accept: application/json' | jq -r '.data.attributes.status')
```
Comment (G-Rath) [inline ci/deploy-to-uat]:
```suggestion
  STATUS=$(
    curl --silent -X GET -u $CWP_DASH_USER:$CWP_DASH_TOKEN $LINK \
      -H 'Content-Type: application/json' -H 'Accept: application/json' \
      | jq -r '.data.attributes.status'
  )
```
🐘 I'd prefer we don't smoosh everything onto a couple of lines - I find it mores readable in this sort of layout.

File: ci/deploy-to-uat [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# This script attempts a git fetch in CWP, waits for it to be ready
+# and creates a deployment request.
+
+# Creates a git fetch request
+curl -X POST -u $CWP_DASH_USER:$CWP_DASH_TOKEN \
+  https://dash.cwp.govt.nz/naut/project/$CWP_STACK_ID/git/fetches \
+  -H 'Content-Type: application/json' -H 'Accept: application/json' >response.json
+
+# We extract the request's status URL from the response
+LINK=$(cat response.json | jq -r '.data.links.self')
```
Comment (G-Rath) [inline ci/deploy-to-uat:20]: 🐘 typically I use lowercase for variables that are not exported/global, to make it more obvious script-local vs env variables.

**Review by G-Rath — CHANGES_REQUESTED** [review]
Could you please run `shellcheck` & `shfmt` over these files? (`prettier-plugin-sh` will run `shfmt`)

File: ci/deploy-to-production (hunk 1 above)
Comment (halles) [reply ci/deploy-to-production:17]: The CWP specific deployment steps need to be fixed in bitbucket first through WCON cause otherwise I won't be able to code and test.

File: ci/deploy-to-uat (hunk 2 above)
Comment (halles) [reply ci/deploy-to-uat]: Odd. I think this was done by prettier or some other format. Either way, same situation as the comment above

Comment (G-Rath) [reply ci/deploy-to-uat]: `prettier-plugin-sh` shouldn't have done that as it's using `shfmt` which doesn't format like this.

I'm not sure why this formatting change is blocked by your above comment, since we're only changing the formatting (which can be verified statically via `shellcheck`).

File: .node-version [hunk 4]
```diff
@@ -0,0 +1 @@
+14
```
Comment (G-Rath) [inline .node-version]: This has to be an exact version of node to work properly - it's probably going to be best to just omit it for now.

In the `rails-template` we generate it via a build step, but for our non-script-generated templates I think we'll need to decide on a common convention to use for this sort of thing.

File: .npm-version [hunk 5]
```diff
@@ -0,0 +1 @@
+6
```
Comment (G-Rath) [inline .npm-version]: I understand why you have this file here, but I think it's going to cause trouble in the long run since it's not actually used by any tool.

Can you please remove it?

File: .composer-version [hunk 6]
```diff
@@ -0,0 +1 @@
+1
```
Comment (G-Rath) [inline .composer-version]: I understand why you have this file here, but I think it's going to cause trouble in the long run since it's not actually used by any tool (afaik - let me know if I'm wrong).

File: ci/build-assets [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+# Echo all commands just before they are executed
+set -x
+
+# Stop processing this script if any of the commands we run return an error
+# (i.e. exit with a non-zero status)
+set -e
+
+pwd
+ls -al public/
+ls -al public/dist/
```
Comment (G-Rath) [inline ci/build-assets]:
```suggestion
```

🐘 these look to me like debugging related commands?

File: README.md [hunk 8]
```diff
@@ -23,39 +23,129 @@ Also refer to our
...
 
-Clone this repository, remove its `.git` directory and initialise it as a new
-repository:
+* Remove this title after setting up the project, to avoid confusion *
 
+Execute the following to create a new project based of this repository, updated
+to the latest version of the SilverStripe stack
+
+```bash
+$ composer create-project --no-install --remove-vcs ackama/silverstripe-cwp-template your-cwp-project
```
Comment (G-Rath) [inline README.md:40]: I forgot that you do this through composer, meaning we could easily have a custom script to generate files if we wanted - but probably best to land that in a followup PR as an improvement; this ones already pretty big 😅

**Review by G-Rath — CHANGES_REQUESTED** [review]
Looks good - I noticed a few places where we could do follow up improvements, but this PR is already large enough so happy for it to be landed as is.

Only change I'd like is to remove the `.<tool>-version` files, and then you should be good to go :)

File: README.md (hunk 8 above)
Comment (halles) [reply README.md:40]: I didn't get this comment 🤔

File: ci/deploy-to-uat (hunk 2 above)
Comment (halles) [reply ci/deploy-to-uat]: It is not blocked but it will be rewritten properly as this piece of code has not been tested or modified at all. I need to get this from WCON when I've ported the pipeline. This is because there's no free CWP tier for us to develop on and will have to use WCON for it.

Comment (halles) [reply ci/deploy-to-uat]: Essentially, whatever is not in `bitbucket-pipeline.yml` can be ported as is, which is why it would be nice to have checked here and moved over. But then in WCON I need to recreate bitbucket-pipeline.yml and the CWP specific deploy steps because that's the only place where we can dev and test. Then those files will come back here. Trying to avoid two giant PRs that are very similar if not the same.

File: .composer-version (hunk 6 above)
Comment (halles) [reply .composer-version]: I though we had agrees to this, but OK. It is documented in the README

File: README.md (hunk 8 above)
Comment (halles) [reply README.md:40]: @G-Rath could you rephrase please? :)

Comment (G-Rath) [reply README.md:40]: This was in relation to [this comment](https://github.com/ackama/silverstripe-cwp-template/pull/20#discussion_r602944302).

Feel free to just ignore it :)

Comment (halles) [reply README.md:40]: @G-Rath gotcha. Yeah, will ignore until we come up with a better idea

**Review by G-Rath — APPROVED** [review]
